File: longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        
        arr = []
        maximum, current = 0, 0
        
        for letter in s:
            if letter in arr:
                current = len(arr)
                index = arr.index(letter)
                arr = arr[index + 1:]
                
            if current > maximum:
                maximum = current
                
            arr.append(letter)
            
        return max(maximum, len(arr))
    
## Solution 2

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        idx_map = {}
        low, current, maximum, last = 0, 0, 0, -1
        
        for i, letter in enumerate(s):
            if letter in idx_map:
                current = i - low
                new_pos = idx_map[letter] + 1
                
                low = new_pos if new_pos > low else low
                
            if current > maximum:
                maximum = current
                
            last = i
            idx_map[letter] = i
            
        return max(maximum, last - low + 1)

File: longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    
    const idxMap = {};
    var [low, current, maximum, last] = [0, 0, 0, -1];
    
    for (let i = 0; i < s.length; i++) {
        let letter = s[i];
        
        if (letter in idxMap) {
            current = i - low;
            let newPos = idxMap[letter] + 1;
            
            low = newPos > low ? newPos: low;
        }
        
        if (current > maximum) {
            maximum = current;
        }
        
        last = i;
        idxMap[letter] = i;
    }
    
    return Math.max(maximum, last - low + 1)
};
